agent/util-scripts/pbench-tool-data-sink

```diff
@@ -0,0 +1,735 @@
+        assert self._tm_tracking is not None, f"Logic bomb! self._tm_tracking is None"
+
+        done = False
+        while not done:
+            for hostname, tm in self._tm_tracking.items():
+                if tm["posted"] == "waiting":
+                    # Don't bother checking any other Tool Meister when we
+                    # have at least one that has not sent any data.
+                    break
+                assert tm["posted"] == "yes", f"Logic bomb! {tm['posted']!r}"
+            else:
+                # We have checked every Tool Meister tracking record and
+                # they all have posted their data (`posted` field is set
+                # to "yes").  So we can safely exit the wait loop
+                done = True
+            if not done:
+                self._cv.wait()
+        return
+
+    def _change_tm_tracking(self, curr, new):
+        """_change_tm_tracking - if we have a tool meister tracking dictionary
+        update the posted state from the current expected value to the target
+        new value.
+
+        No changes take place if the tool meister tracking dictionary does not
+        exist yet.
+
+        Assumes self._lock is already acquired by our caller.
+        """
+        if self._tm_tracking is None:
+            return
+        for hostname, tm in self._tm_tracking.items():
+            assert (
+                tm["posted"] is None or tm["posted"] == curr
+            ), f"_change_tm_tracking unexpected tm posted value, {tm!r}"
+            tm["posted"] = new
+
+    def state_change(self, data):
+        """state_change - give a data dictionary, change the state for this
+        data sink instance.
+
+        The "watcher" thread has already validated the state field, we then
+        validate the directory field.
+
+        Public method, returns None, raises no exceptions explicitly, called
+        by the "watcher" thread.
+        """
+        if self.state is None:
+            # This is the first published state change we have received.
+            with self._lock:
+                assert self._tm_tracking is None, (
+                    f"Logic bomb! self._tm_tracking '{self._tm_tracking!r}'"
+                    " is not None"
+                )
+                # Typically, we only call this method once when the Tool Data
+                # Sink first starts since the list of Tool Meister's is
+                # static.  We call it when we get the first message to
+                # be sure the "tm-pids" key exists.
+                #
+                # FIXME: what happens when a Tool Meister dies, and rejoins?
+                self._tm_tracking = self._fetch_tms()
+
+        self._data = data
+        self.state = data["state"]
+        if self.state == "terminate":
+            raise self.Terminate("Terminate bottle server")
+        directory_str = data["directory"]
+        directory = Path(directory_str)
```

Comment:
It was intended that `directory` would be the `Path` variable, and `directory_str` would be the raw string taken from the payload, so that later we don't have to fetch it from `data`.  So this makes it explicit.